### PR TITLE
Add BigTable source format in BigQuery tables

### PR DESCRIPTION
**Pull request by jmthvt**
This is adding the BIGTABLE source format in BigQuery tables. 
Fixes https://github.com/hashicorp/terraform-provider-google/issues/7653

If this PR is for Terraform, I acknowledge that I have:

- [x] Searched through the [issue tracker](https://github.com/hashicorp/terraform-provider-google/issues) for an open issue that this either resolves or contributes to, commented on it to claim it, and written "fixes {url}" or "part of {url}" in this PR description. If there were no relevant open issues, I opened one and commented that I would like to work on it (not necessary for very small changes).
- [x] [Generated Terraform](https://github.com/GoogleCloudPlatform/magic-modules#generating-downstream-tools), and ran `make test` and `make lint` to ensure it passes unit and linter tests.
- [x] Ensured that all new fields I added that can be set by a user appear in at least one [example](https://github.com/GoogleCloudPlatform/magic-modules/tree/master/templates/terraform/examples) (for generated resources) or [third_party test](https://github.com/GoogleCloudPlatform/magic-modules/tree/master/third_party/terraform/tests) (for handwritten resources or update tests).
- [x] [Ran](https://github.com/hashicorp/terraform-provider-google/blob/master/.github/CONTRIBUTING.md#tests) relevant acceptance tests (If the acceptance tests do not yet pass or you are unable to run them, please let your reviewer know).
- [x] Read the [Release Notes Guide](https://github.com/GoogleCloudPlatform/magic-modules/blob/master/.ci/RELEASE_NOTES_GUIDE.md) before writing my release note below.

**Release Note Template for Downstream PRs (will be copied)**

```release-note:enhancement
bigquery: Added BigTable source format in BigQuery table
```

### CODE REVIEW

File: third_party/terraform/tests/resource_bigquery_table_test.go [hunk 1]
```diff
@@ -416,6 +416,30 @@ func TestAccBigQueryDataTable_sheet(t *testing.T) {
...
+		"random_suffix": randString(t, 10),
+	}
+
+	vcrTest(t, resource.TestCase{
+		PreCheck:     func() { testAccPreCheck(t) },
+		Providers:    testAccProviders,
+		CheckDestroy: testAccCheckBigQueryTableDestroyProducer(t),
+		Steps: []resource.TestStep{
+			{
+				Config: testAccBigQueryTableFromBigtable(context),
```
Comment (slevenick) [inline third_party/terraform/tests/resource_bigquery_table_test.go:432]: Looks like this doesn't match the function declaration, causing this error:
`google-beta/resource_bigquery_table_test.go:432:13: undefined: testAccBigQueryTableFromBigtable`

**Review by slevenick — CHANGES_REQUESTED** [review]

File: third_party/terraform/tests/resource_bigquery_table_test.go (hunk 1 above)
Comment (jmthvt) [reply third_party/terraform/tests/resource_bigquery_table_test.go:432]: oops, thanks!
Should be fixed now.

File: third_party/terraform/tests/resource_bigquery_table_test.go [hunk 2]
```diff
@@ -1048,6 +1072,37 @@ func testAccBigQueryTableFromSheet(context map[string]interface{}) string {
...
+		dataset_id = google_bigquery_dataset.dataset.dataset_id
+		table_id   = "tf_test_bigtable_%{random_suffix}"
+
+		external_data_configuration {
+		  autodetect            = true
+		  source_format         = "BIGTABLE"
+		  ignore_unknown_values = true
+
+		  source_uris = [
+			"https://googleapis.com/bigtable/projects/project_id/instances/instance_id/tables/table_name",
```
Comment (slevenick) [inline third_party/terraform/tests/resource_bigquery_table_test.go:1087]: Looks like this is causing an error during testing. This is an invalid source, maybe try using one of the source_uris from another test?

File: third_party/terraform/tests/resource_bigquery_table_test.go [hunk 3]
```diff
@@ -416,6 +416,30 @@ func TestAccBigQueryDataTable_sheet(t *testing.T) {
 	})
 }
 
+func TestAccBigQueryDataTable_bigtable(t *testing.T) {
```
Comment (slevenick) [inline third_party/terraform/tests/resource_bigquery_table_test.go:419]: This test should only be run for the `beta` provider, right? This file may need to be renamed to end in `.go.erb` and wrap this test in if/else tags to check the version.

An example can be seen here: https://github.com/GoogleCloudPlatform/magic-modules/blob/master/third_party/terraform/tests/resource_binaryauthorization_policy_test.go.erb#L41

**Review by slevenick — CHANGES_REQUESTED** [review]

File: third_party/terraform/tests/resource_bigquery_table_test.go (hunk 3 above)
Comment (jmthvt) [reply third_party/terraform/tests/resource_bigquery_table_test.go:419]: That's quite a weird one actually:
You can set the format to BIGTABLE in stable (search for sourceFormat in https://bigquery.googleapis.com/discovery/v1/apis/bigquery/v2/rest )
However only 2 regions are supported for the queries https://cloud.google.com/bigquery/external-data-bigtable#supported_regions_and_zones.

From the BigQuery API point of view, I don't think it requires the beta provider.
But from BigTable it does.
Not sure of the preferred approach here.

File: third_party/terraform/tests/resource_bigquery_table_test.go (hunk 2 above)
Comment (jmthvt) [reply third_party/terraform/tests/resource_bigquery_table_test.go:1087]: I'll try to create a bigtable in the test & generate the URI.

Comment (jmthvt) [reply third_party/terraform/tests/resource_bigquery_table_test.go]: @slevenick that's added but I'm unable to run the tests - can you please try again?

File: third_party/terraform/resources/resource_bigquery_table.go [hunk 4]
```diff
@@ -153,9 +153,9 @@ func resourceBigQueryTable() *schema.Resource {
 						"source_format": {
 							Type:        schema.TypeString,
 							Required:    true,
-							Description: `The data format. Supported values are: "CSV", "GOOGLE_SHEETS", "NEWLINE_DELIMITED_JSON", "AVRO", "PARQUET", and "DATASTORE_BACKUP". To use "GOOGLE_SHEETS" the scopes must include "googleapis.com/auth/drive.readonly".`,
+							Description: `The data format. Supported values are: "CSV", "GOOGLE_SHEETS", "NEWLINE_DELIMITED_JSON", "AVRO", "PARQUET", "DATSTORE_BACKUP", and "BIGTABLE". To use "GOOGLE_SHEETS" the scopes must include "googleapis.com/auth/drive.readonly".`,
```
Comment (venkykuberan) [inline third_party/terraform/resources/resource_bigquery_table.go:156]: I wonder `ORC` can also included in the source_format list ?. We have an open issue (https://github.com/hashicorp/terraform-provider-google/issues/7691) for that type.

Comment (jmthvt) [reply third_party/terraform/resources/resource_bigquery_table.go:156]: Hi, sorry but that's out of scope of this PR - adding it in the list is straightforward but requires to write the tests.

**Comment by jmthvt** [comment]
@slevenick are the tests passing?

File: third_party/terraform/tests/resource_bigquery_table_test.go [hunk 5]
```diff
@@ -1048,6 +1073,62 @@ func testAccBigQueryTableFromSheet(context map[string]interface{}) string {
...
+		instance_type = "DEVELOPMENT"
+		deletion_protection = false
+	}
+
+	resource "google_bigtable_table" "table" {
+		name          = "%{random_suffix}"
+		instance_name = google_bigtable_instance.instance.name
+
+		column_family {
+			family = cf-"%{random_suffix}-first"
```
Comment (slevenick) [inline third_party/terraform/tests/resource_bigquery_table_test.go:1095]: Looks like this is incorrect. Quotes should wrap the `cf-`.

```
Error: Missing newline after argument
on config893095785/terraform_plugin_test.tf line 19, in resource "google_bigtable_table" "table":
19:       family = cf-"j42ngv1ayb-first"
An argument definition must end with a newline.
```

**Review by slevenick — CHANGES_REQUESTED** [review]

File: third_party/terraform/tests/resource_bigquery_table_test.go (hunk 5 above)
Comment (jmthvt) [reply third_party/terraform/tests/resource_bigquery_table_test.go]: sorry for that - fixed now.

**Comment by jmthvt** [comment]
Hey @slevenick, sorry to chase, are the tests passing?

**Comment by slevenick** [comment]
> Hey @slevenick, sorry to chase, are the tests passing?

Sorry, lost track of this. Looks like they aren't passing:
```
TestAccBigQueryDataTable_bigtable: provider_test.go:266: Step 1/2 error: Error running apply: 2020/11/13 19:39:42 [DEBUG] Using modified User-Agent: Terraform/0.12.29 HashiCorp-terraform-exec/0.10.0
Error: Error creating instance. rpc error: code = InvalidArgument desc = Error in field 'instance_id' : Invalid id for collection instances : Length should be between [6,33], but found 36 'tf_test_bigtable_instance_ur4ywgovou'
```

Looks like the name of the instance is too long. Should be an easy fix. You'll also need to rebase or merge master in to resolve conflicts

**Comment by jmthvt** [comment]
@slevenick Updated, can you please try again and let me know. Fingers crossed 🤞

**Comment by slevenick** [comment]
Looks like it no longer compiles:
```
google-beta/resource_bigquery_table.go:166:20: syntax error: unexpected :, expecting comma or )
google-beta/resource_bigquery_table.go:642:2: syntax error: unexpected }, expecting expression
```

**Comment by jmthvt** [comment]
Sorry, introduced this when I merged master. Fixed now.

**Comment by jmthvt** [comment]
@slevenick 👀
Sorry to bother really. If there is a way for me to look at those tests, please let me know!

**Comment by slevenick** [comment]
> @slevenick 👀
> Sorry to bother really. If there is a way for me to look at those tests, please let me know!

```Error: Error creating instance. rpc error: code = InvalidArgument desc = Error in field 'instance_id' : Invalid id for collection instances : Should match [a-z][a-z0-9\-]+[a-z0-9] but found 'tf_test_bigtable_inst_aq1wv0hf8g'```

Looks like that the `instance_id` is in the wrong format. Are you able to compile magic modules? It should produce a provider that you can then run the tests on: https://github.com/GoogleCloudPlatform/magic-modules#testing-your-changes

**Comment by slevenick** [comment]
It looks like there are some bad merge issues going on here with the tests. The test file shows a lot of changes that I don't think are associated with your PR. Can you fix that before I can review this again?

**Comment by jmthvt** [comment]
Yes I will, I've noticed the bad diff too.
Sorry for letting the PR rot, I'm resuming this, set up a sandbox and let you know once ready.

**Comment by jmthvt** [comment]
@slevenick _Finally_ tests are passing locally, this is now ready to review.
My apologies for the chaotic PR and thanks a lot for your patience.

**Review by slevenick — APPROVED** [review]
Looks reasonable, thanks!

**Pull request merged**